### ISSUE TITLE
kube-proxy should periodically resync iptables to recover from unexpected drift

### ISSUE THREAD

**Issue by wedaly** [kubernetes/kubernetes#129128]
### What would you like to be added?

kube-proxy performs a periodic full resync of iptables to correct unexpected drift between expected and actual iptables state.

### Why is this needed?

Today, kube-proxy performs a full resync of iptables only under certain conditions. In AKS, we have sometimes seen cases where iptables rules are accidentally modified (e.g. by daemonsets installed by a customer). Without some trigger (like a restart or endpoints created/deleted), kube-proxy won't detect or correct the drift between the actual/expected state of iptables.

A specific scenario in which this can happen:
1. iptables monitor polls until the canary chain in the mangle table is flushed, but doesn't check the canary chains in NAT or filter table: https://github.com/kubernetes/kubernetes/blob/d001d5684e69b08e120dc328917925318c5f324c/pkg/util/iptables/iptables.go#L586
2. Someone accidentally flushes the NAT table (`iptables -t nat -X`), but not the mangle table.

**Comment by wedaly** [comment]
Discussed in Slack: https://kubernetes.slack.com/archives/C09QYUH5W/p1730303453175959

The current assumption in kube-proxy is that people should avoid modifying iptables in a way that breaks kube-proxy.

This issue is to discuss whether it's worth adding a periodic resync (say once an hour) to ensure that kube-proxy self-recovers eventually if this assumption is violated.

**Comment by aojea** [comment]
This is an interesting topic, so kube-proxy reconciles based on events triggered by Services and Endpoints... 

iptables mode also reconciles on an event triggered by a missing canary chain, but basically because as Dan explains in the slack thread iptables maintain a single global state for all rules and chains, so it is not uncommon to have unintended interactions between different process modifying the rules. nftables allows for per-table state, meaning each table has its own independent state that reduces the risk of unintended interactions.

Checking the code I can see

https://github.com/kubernetes/kubernetes/blob/d001d5684e69b08e120dc328917925318c5f324c/pkg/proxy/iptables/proxier.go#L325-L328

and 

https://github.com/kubernetes/kubernetes/blob/d001d5684e69b08e120dc328917925318c5f324c/pkg/proxy/nftables/proxier.go#L279-L280

is not NewBoundedFrequencyRunner already running after 1 hour if nothing happens?
if is running, is the problem we are not doing a full sync during this execution?

**Comment by wedaly** [comment]
> is not NewBoundedFrequencyRunner already running after 1 hour if nothing happens?
if is running, is the problem we are not doing a full sync during this execution?

Yes, exactly. `proxy.syncRunner` invokes `proxier.syncProxyRules` every 1hr. However, `syncProxyRules` won't perform a full sync unless `proxier.needFullSync` is true
https://github.com/kubernetes/kubernetes/blob/a499facee693a1a83daadb82d88f7b51d324ffc5/pkg/proxy/iptables/proxier.go#L811

In contrast, `ipt.Monitor` calls `proxier.forceSyncProxyRules` when it detects that the mangle canary chain has been deleted
https://github.com/kubernetes/kubernetes/blob/a499facee693a1a83daadb82d88f7b51d324ffc5/pkg/proxy/iptables/proxier.go#L330-L331

which does a full sync
https://github.com/kubernetes/kubernetes/blob/a499facee693a1a83daadb82d88f7b51d324ffc5/pkg/proxy/iptables/proxier.go#L788-L794

**Comment by danwinship** [comment]
Yeah, so the history is:

- pre-[`iptables.Monitor`](https://github.com/kubernetes/kubernetes/pull/81517):
  - kube-proxy queues a complete ruleset rewrite:
    - at startup as soon as it has synced with the apiserver
    - within 1 second of any Service or endpoint changing
    - within 1 second of the node topology labels changing
    - at least once every 30 seconds even if nothing changes.

- post-`iptables.Monitor`, pre-[`MinimizeIPTablesRestore`](https://github.com/kubernetes/kubernetes/pull/110268):
  - kube-proxy queues a complete ruleset rewrite:
    - at startup as soon as it has synced with the apiserver
    - within 1 second of any Service or endpoint changing
    - within 1 second of the node topology labels changing
    - within 30 seconds of a complete iptables ruleset flush
    - at least once an hour even if nothing changes.

- post-`MinimizeIPTablesRestore`:
  - kube-proxy queues a complete ruleset rewrite:
    - at startup as soon as it has synced with the apiserver
    - within 1 second of the node topology labels changing
    - within 30 seconds of a complete iptables ruleset flush
  - kube-proxy queues a partial sync for only changed Services and endpoints:
    - within 1 second of any Service or endpoint changing
    - at least once an hour even if nothing changes.

(where "1 second" means "`MinSyncPeriod`", "30 seconds" means "`SyncPeriod`", and "one hour" is not configurable.)

So the "at least once an hour" sync is now mostly a no-op. It should have ended up in the "rewrites every rule" pile rather than the "rewrites changed rules" pile, but it didn't.

**Comment by aojea** [comment]
> So the "at least once an hour" sync is now mostly a no-op. It should have ended up in the "rewrites every rule" pile rather than the "rewrites changed rules" pile, but it didn't.

I think is reasonable to expect that the after one hour of inactivity we do a full resync, I defer to Dan to ack to accept this feature, I'm +1 on changing this behavior to do a full sync after the hour

**Comment by danwinship** [comment]
oh, yeah, I definitely think we should fix this. I didn't notice it was still needs-triage

/triage accepted

**Comment by olderTaoist** [comment]
can let me try? I raised an [issue](https://github.com/kubernetes/kubernetes/issues/128879) about iptables mode  a while ago, and I want to learn more about kube-proxy iptables mode based on this issue.

**Comment by danwinship** [comment]
@olderTaoist Sure! The Kubernetes Contributor website has a bunch of [docs on how to get started](https://www.kubernetes.dev/docs/); if you have more specific questions on where to start or what approach to take for this particular bug, you can ask here.

**Comment by aroradaman** [comment]
We can experiment with netlink to see if the canary chain exists in all the tables, that shouldn't have any performance penalty.

I couldn't find any API to read iptables rules, maybe we can use nftables netlink API to read rules created by iptables-nft.
@danwinship @aojea  We can experiment with https://github.com/google/nftables

**Comment by aojea** [comment]
nftables can monitor changes on the objects https://wiki.nftables.org/wiki-nftables/index.php/Monitoring_ruleset_updates

**Comment by danwinship** [comment]
I don't think we want to make the iptables backend _require_ iptables-nft rather than iptables-legacy.

FWIW recent kernels introduce the `owned` and `persistent` flags for nftables tables, which makes it so that the table is owned by a single process and no other processes can modify it. So eventually we'll be able to depend on that in the nftables backend.
